File: java/BinaryTreeDiameter.java
class Solution {

    /**
     * Function that evaluates the diameter of a binary tree.
     * @param root the starting node of the tree.
     * @return the diamater of the tree.
     */
    public int diameterOfBinaryTree(TreeNode root) {

        //Base case null/empty tree.
        if(root == null){
            return 0;
        }

        /**
         * Get the left and right heights.
         */
        int leftHeight = height(root.left);
        int rightHeight = height(root.right);

        /**
         * Call algo recursively over the left and right subtrees
         */
        int leftDiam = diameterOfBinaryTree(root.left);
        int rightDiam = diameterOfBinaryTree(root.right);

        return Math.max(leftHeight + rightHeight, Math.max(leftDiam, rightDiam));
    }

    /**
     * Helper method used to calculate the height of the tree.
     * Returns the longest height because we want the longest path between any two nodes.
     */
    public int height(TreeNode node){
        if(node == null){
            return 0;
        }
        return (1 + Math.max(height(node.left), height(node.right)));
    }
}

File: python/BinaryTreeDiameter.py
from typing import Optional
from python.TreeNode import TreeNode

class Solution:

    # The diamater of the tree is the longest path between any two nodes, *does not* necessarily include the root. 
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:

        if root is None:
            return 0
        
        # Take the left and right heights of the trees. 
        left_height = self.height(root.left)
        right_height = self.height(root.right)

        # Given that the longest path might not pass through the node, create left and right diameter and call them recursively.
        left_diam = self.diameterOfBinaryTree(root.left)
        right_diam = self.diameterOfBinaryTree(root.right)

        # Return the max between: 1. sum of left and right heights (longest path passes through node) 2. the max of the left and right diameter (path does not pass through root)
        return max(left_height + right_height, max(left_diam, right_diam))


    # Basic height calculation method. 
    def height(self, node: Optional[TreeNode]) -> int:

        # Base case, null root. 
        if node is None:
            return 0
        
        left_height = self.height(node.left)
        right_height = self.height(node.right)

        return max(left_height, right_height) + 1